### PR TITLE
Add support for timing C++ snippets.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47865 Enable callgrind collection for C++ snippets
* **#47864 Add support for timing C++ snippets.**
* #47863 Rework compat bindings.
* #47862 Add option for cpp_extensions to compile standalone executable
* #47861 Expose CXX_FLAGS through __config__

Differential Revision: [D25199262](https://our.internmc.facebook.com/intern/diff/D25199262)

### CODE REVIEW

File: torch/utils/benchmark/utils/timer.py [hunk 1]
```diff
@@ -158,15 +210,15 @@ def __init__(
         # Stripping this down to 'print("This is a stmt")' doesn't change
         # what gets executed, but it makes __repr__'s nicer.
         stmt = textwrap.dedent(stmt)
-        stmt = (stmt[1:] if stmt[0] == "\n" else stmt).rstrip()
+        stmt = (stmt[1:] if stmt and stmt[0] == "\n" else stmt).rstrip()
         setup = textwrap.dedent(setup)
-        setup = (setup[1:] if setup[0] == "\n" else setup).rstrip()
+        setup = (setup[1:] if setup and setup[0] == "\n" else setup).rstrip()
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:215]: I agree the new code is better than the old code, but this is my skeptical face 🤔

File: torch/utils/benchmark/utils/cpp_jit.py [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
-from torch.utils.benchmark.utils._stubs import CallgrindModuleType
+from torch.utils.benchmark.utils._stubs import CallgrindModuleType, TimeitModuleType
 from torch.utils import cpp_extension
 
 
 LOCK = threading.Lock()
 SOURCE_ROOT = os.path.split(os.path.abspath(__file__))[0]
+BUILD_ROOT = os.path.join(
+    torch._appdirs.user_cache_dir(appname="benchmark_utils_jit"),
+    f"build_{uuid.uuid4()}".replace("-", "")
```
Comment (ezyang) [inline torch/utils/benchmark/utils/cpp_jit.py:20]: Calcing the build uuid only once at module import seems pretty skeevy to me...

File: test/benchmark_utils/test_benchmark_utils.py [hunk 3]
```diff
@@ -162,6 +162,16 @@ def test_timer(self):
...
         self.assertIsInstance(sample, float)
 
+    @slowTest
+    def test_cpp_timer(self):
+        timer = benchmark_utils.Timer(
+            "torch::Tensor y = x + 1;",
+            setup="torch::Tensor x = torch::empty({1});",
+            language=benchmark_utils.Language.CPP,
+        )
+        t = timer.timeit(10)
```
Comment (ezyang) [inline test/benchmark_utils/test_benchmark_utils.py:172]: Is there really any good reason to run this for 10s? For the purpose of a test haha

File: torch/utils/benchmark/utils/common.py [hunk 4]
```diff
@@ -48,7 +48,7 @@ def title(self) -> str:
 
     def setup_str(self) -> str:
         return (
-            "" if self.setup == "pass"
+            "" if (self.setup == "pass" or not self.setup)
```
[the remark this answers is not in the document]
Comment (robieta) [reply torch/utils/benchmark/utils/common.py:51]: I don't think I can, since it's a user defined string. It's kind of like an `Optional[str]`, except that `timeit.Timer` doesn't allow an empty `setup`. I agree it's slightly curious.

File: torch/utils/benchmark/utils/cpp_jit.py (hunk 2 above)
Comment (robieta) [reply torch/utils/benchmark/utils/cpp_jit.py:34]: I was about to write a big paragraph on why it's not, but then I thought "that big paragraph should be a comment!".

File: test/benchmark_utils/test_benchmark_utils.py (hunk 3 above)
Comment (robieta) [reply test/benchmark_utils/test_benchmark_utils.py:172]: It's 10 iterations. `timer.blocked_autorange(min_run_time=10)` would be 10 seconds.

Comment (ezyang) [reply test/benchmark_utils/test_benchmark_utils.py:172]: Converse question, then, is why is this a slow test :) (compile time?)

File: torch/utils/benchmark/utils/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/utils/benchmark/utils/common.py:51]: oh, pass isn't a keyword, it's literally the Python code "pass". That's curious, but I guess matching timeit.Timer takes precedence here.

File: torch/utils/benchmark/utils/cpp_jit.py (hunk 2 above)
Comment (ezyang) [reply torch/utils/benchmark/utils/cpp_jit.py:34]: Hmm... well this makes sense, but it raises a new question: do you *trust* `cpp_extension` s recompilation cache? When it comes to benchmarking, I'm not sure I would...

File: test/benchmark_utils/test_benchmark_utils.py (hunk 3 above)
Comment (robieta) [reply test/benchmark_utils/test_benchmark_utils.py:172]: Yup, compile time.

File: torch/utils/benchmark/utils/timer.py [hunk 5]
```diff
@@ -135,21 +176,32 @@ def __init__(
...
-        # We copy `globals` to prevent mutations from leaking, (for instance,
-        # `eval` adds the `__builtins__` key) and include `torch` if not
-        # specified as a convenience feature.
-        globals = dict(globals or {})
-        globals.setdefault("torch", torch)
-        self._globals = globals
+        # We copy `globals` to prevent mutations from leaking.
+        # (For instance, `eval` adds the `__builtins__` key)
+        self._globals = dict(globals or {})
+        if language in (Language.PYTHON, "py", "python"):
```
Comment (ezyang) [inline torch/utils/benchmark/utils/timer.py:187]: nit: would be cleaner to have a function that takes care of normalizing string to the Language Enum, and then use that. I don't think it's a big deal in this patch because the logic here is still centralized, but the moment you want to have a second interface that takes in string languages, you're gonna need this refactor.

File: torch/utils/benchmark/utils/cpp_jit.py [hunk 6]
```diff
@@ -66,3 +89,55 @@ def get_compat_bindings() -> CallgrindModuleType:
...
+            src
+        )
+
+    # We want to isolate different Timers. However `cpp_extension` will
+    # cache builds which will significantly reduce the cost of repeated
+    # invocations.
+    with LOCK:
+        if not os.path.exists(BUILD_ROOT):
+            os.makedirs(BUILD_ROOT)
+            atexit.register(shutil.rmtree, BUILD_ROOT)
```
Comment (ezyang) [inline torch/utils/benchmark/utils/cpp_jit.py:114]: https://github.com/pytorch/pytorch/pull/47428/files#r521094904

**Review by ezyang — APPROVED** [review]
okey dokey

File: torch/utils/benchmark/utils/cpp_jit.py (hunk 2 above)
Comment (robieta) [reply torch/utils/benchmark/utils/cpp_jit.py:34]: I have two safeguards:
1) The `uuid` in the build root, so it won't mix `libtorch` across PyTorch versions.
2) I put the hash of the source in the name before I pass it to `cpp_extension`

So really `cpp_jit.py` guarantees cache correctness; it just has to make sure it doesn't also vary something that `cpp_extension` caches on. (i.e. problems will manifest as excessive recompilation, but never correctness issues.)

Want me to add a note to that effect?

File: torch/utils/benchmark/utils/cpp_jit.py (hunk 6 above)
Comment (robieta) [reply torch/utils/benchmark/utils/cpp_jit.py:114]: I'm implicitly relying on `torch._appdirs.user_cache_dir` having similar kinds of litter safety. (That's what you would get from vanilla `cpp_extension` anyway) On my machine it's `~/.cache/...`, for instance. And if you're still unconvinced it's no trouble to change it to use `tempfile` instead.

Comment (ezyang) [reply torch/utils/benchmark/utils/cpp_jit.py:114]: https://askubuntu.com/questions/102046/is-it-okay-to-delete-the-cache-folder seems to suggest that `~/.cache` doesn't get automatically cleaned.
